### PR TITLE
Dupe fixes and more robust backpacks and pouches

### LINKED ISSUE #307: [Triffles] Pouches can dupe themselves inside other pouches
*Forgot to say, The Server is Paper 1.21.4

When Opening a pouch you, usually you cant put other pouches into it BUT you can bypass this by having a pouch in your hotbar and press the corresponding number (on which contains the pouch) while hovering over an spot in the pouch and even put itself inside of it.

While this seems to be just a trivial thing to give yourself extra storage by having a pouch hold other pouches or even lose the pouch if you leave itself inside of itself, the issue comes when you do the same method but immediately take the pouch out. While you may have the pouch you took out in your inventory as normal, when you open the main pouch (the one you forced in another pouch into but quickly took out) you will see the pouch is duped with all items it had before in it.

The photos are of a case where you put the pouch inside of itself, take it out and now you have 2 pouches of the same items:

First photo is the pouch in the hotbar and I put it inside the empty slow with the number 3,
Second of it being inside of itself
Third is Showing the pouch is back to the 3rd Hotbar slot BUT the pouch is also inside, which you can take out and reap the rewards

Hopefully I Explained This Well! Just be aware of this

Image

Image

Image

### PR DESCRIPTION
**Pull request by FaultyFunctions**
This fixes #307, #314. I know there are already a few open PRs about this (closes #319, #317, #315) and I don't mean to step on any toes, but I feel this is a more robust way to fix the issues that plague both pouches and backpacks.

What this PR does:
* Fixes pouch/backpack dupes that were caused by moving the pouch/backpacks while they were open. It does this by finding the lost ItemStack reference by checking a data tag on all storage items in the player's inventory.
* Allows dropping opened pouches/backpacks without fear of dupes. The inventory will close accordingly if the player drops the currently opened pouch/backpack.
* Prevents hotbar/number key swapping storage items into the storage inventory, but allows it in the player's inventory.
* Allows the player to move both pouches and backpacks freely within the player inventory without data loss/dupes.
* Fixes any old pouches/backpacks that don't have the `vane_trifles:currently_opened_storage` data tag by adding it upon opening the pouch/backpack.

### CODE REVIEW

**Comment by FaultyFunctions** [comment]
I tested this on a backup of my server and I haven't ran into any issues. Dupes are gone and player's old pouches and backpacks work well.

File: vane-trifles/src/main/java/org/oddlama/vane/trifles/StorageGroup.java [hunk 1]
```diff
@@ -138,11 +149,56 @@ public void on_inventory_click(InventoryClickEvent event) {
...
+        }
+
+        if (!is_storage_item(event.getItem().getItemStack())) {
+            return;
+        }
+
+        // Ensure bugged/old storage items are set to closed when picked up just in case
+        event.getItem().getItemStack().editMeta(meta -> {
+            meta.getPersistentDataContainer().set(STORAGE_IS_OPEN, PersistentDataType.BOOLEAN, false);
+        });
```
Comment (oddlama) [inline vane-trifles/src/main/java/org/oddlama/vane/trifles/StorageGroup.java:199]: Very nice, I was just about to ask what happens when the client disconnects while having a storage item opened, or when a server crashes occurs.

Altough I'm not entirely sure whether it will be possible to move or drop bugged items at all times.

**Review by oddlama — COMMENTED** [review]
Looks pretty great, thanks for your efforts! I'll test this on my test instance tomorrow and then decide how to proceed :)

**Comment by oddlama** [comment]
Just tested this and it seems to works very well! I didn't find any edge cases yet (with my limited testing :P). I think we can merge this and respond when/if anything pops up in the future. Thanks to all of you for your efforts!
